// @parcel/core/lib/ContentGraph.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _Graph() {
  const data = _interopRequireDefault(require("./Graph"));

  _Graph = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ContentGraph extends _Graph().default {
  constructor(opts) {
    if (opts) {
      let {
        _contentKeyToNodeId,
        ...rest
      } = opts;
      super(rest);

      _defineProperty(this, "_contentKeyToNodeId", void 0);

      this._contentKeyToNodeId = _contentKeyToNodeId;
    } else {
      super();

      _defineProperty(this, "_contentKeyToNodeId", void 0);

      this._contentKeyToNodeId = new Map();
    }
  } // $FlowFixMe[prop-missing]


  static deserialize(opts) {
    return new ContentGraph(opts);
  } // $FlowFixMe[prop-missing]


  serialize() {
    return { ...super.serialize(),
      _contentKeyToNodeId: this._contentKeyToNodeId
    };
  }

  addNodeByContentKey(contentKey, node) {
    if (!this.hasContentKey(contentKey)) {
      let nodeId = super.addNode(node);

      this._contentKeyToNodeId.set(contentKey, nodeId);

      return nodeId;
    } else {
      let existingNodeId = this.getNodeIdByContentKey(contentKey);
      let existingNode = (0, _nullthrows().default)(this.getNodeByContentKey(contentKey));
      existingNode.value = node.value;
      this.updateNode(existingNodeId, existingNode);
      return existingNodeId;
    }
  }

  getNodeByContentKey(contentKey) {
    let nodeId = this._contentKeyToNodeId.get(contentKey);

    if (nodeId != null) {
      return super.getNode(nodeId);
    }
  }

  getNodeIdByContentKey(contentKey) {
    return (0, _nullthrows().default)(this._contentKeyToNodeId.get(contentKey), 'Expected content key to exist');
  }

  hasContentKey(contentKey) {
    return this._contentKeyToNodeId.has(contentKey);
  }

  removeNode(nodeId) {
    this._assertHasNodeId(nodeId);

    this._contentKeyToNodeId.delete((0, _nullthrows().default)(this.getNode(nodeId)).id);

    super.removeNode(nodeId);
  }

}

exports.default = ContentGraph;
