// @parcel/core/lib/Dependency.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDependency = createDependency;
exports.mergeDependencies = mergeDependencies;

function _hash() {
  const data = require("@parcel/hash");

  _hash = function () {
    return data;
  };

  return data;
}

function createDependency(opts) {
  var _opts$sourceAssetId, _opts$pipeline, _opts$isAsync, _opts$isOptional, _opts$isURL, _opts$isIsolated;

  let id = opts.id || (0, _hash().hashString)(((_opts$sourceAssetId = opts.sourceAssetId) !== null && _opts$sourceAssetId !== void 0 ? _opts$sourceAssetId : '') + opts.moduleSpecifier + opts.env.id + (opts.target ? JSON.stringify(opts.target) : '') + ((_opts$pipeline = opts.pipeline) !== null && _opts$pipeline !== void 0 ? _opts$pipeline : ''));
  return { ...opts,
    id,
    isAsync: (_opts$isAsync = opts.isAsync) !== null && _opts$isAsync !== void 0 ? _opts$isAsync : false,
    isEntry: opts.isEntry,
    isOptional: (_opts$isOptional = opts.isOptional) !== null && _opts$isOptional !== void 0 ? _opts$isOptional : false,
    isURL: (_opts$isURL = opts.isURL) !== null && _opts$isURL !== void 0 ? _opts$isURL : false,
    isIsolated: (_opts$isIsolated = opts.isIsolated) !== null && _opts$isIsolated !== void 0 ? _opts$isIsolated : false,
    meta: opts.meta || {},
    symbols: opts.symbols
  };
}

function mergeDependencies(a, b) {
  let {
    meta,
    symbols,
    ...other
  } = b;
  Object.assign(a, other);
  Object.assign(a.meta, meta);

  if (a.symbols && symbols) {
    for (let [k, v] of symbols) {
      a.symbols.set(k, v);
    }
  }
}
